core/src/ldml/ldml_processor.cpp

```diff
@@ -476,7 +476,9 @@ ldml_event_state::emit_text(const std::u32string &str) {
 
 void
 ldml_event_state::emit_text( km_core_usv ch) {
-  assert(ch != LDML_UC_SENTINEL);
+  // assert that no noncharacter gets through (which includes UC_SENTINEL)
+  // markers will go through emit_marker() and not through this function.
+  assert(km::core::kmx::Uni_IsValid(ch));
```

Comment:
```suggestion
  assert(km::core::kmx::Uni_IsValid(ch));
  if(!km::core::kmx::Uni_IsValid(ch)) return;
```

`assert` is only present in debug builds, so we should probably fall out in release builds, or else, if that would break other assumptions on text length etc, replace ch with a safe char:

```suggestion
  assert(km::core::kmx::Uni_IsValid(ch));
  if(!km::core::kmx::Uni_IsValid(ch)) ch = 0x0020; // release build: replace invalid with space character
```

I should have caught this when we wrote the original assertion.

Reply:
@srl295 wrote:

> Not FFFD?

I'm not sure. It's an error condition so FFFD would be a good choice. I was opting for principle of least irritation to end user but this happening at all probably means bad things gone wrong internally so FFFD also stops further crashes and lets the user know that something is wrong.

Reply:
U+FFFD  � would be clearer that something bad happened rather than silently failing - or spacing out.  

This PR also evaluates every string coming in— although, I just thought of a case i need to check also. Anyway, it _will_ evaluate every string, so yes this should be something very strange if it made it to emit.

Reply:
Yeah, let's go with U+FFFD then. If users see that coming up in their documents, they are likely to complain, whereas if they see space they will probably just be confused.

Reply:
```suggestion
  assert(km::core::kmx::Uni_IsValid(ch));
  if(!km::core::kmx::Uni_IsValid(ch)) ch = 0xFFFD; // release build: replace invalid with � to indicate something's gone badly wrong.
```